### PR TITLE
Fix ReadOnlyEngineTests#testSearcherId

### LINKED ISSUE #66869: ReadOnlyEngineTests testSearcherId fails
**Build scan**:

https://gradle-enterprise.elastic.co/s/j4tztllphmnpo

**Repro line**:
./gradlew ':server:test' --tests "org.elasticsearch.index.engine.ReadOnlyEngineTests.testSearcherId" \
  -Dtests.seed=5F23C74BF9BA02D2 \
  -Dtests.security.manager=true \
  -Dtests.locale=es-AR \
  -Dtests.timezone=America/Noronha \
  -Druntime.java=11

**Reproduces locally?**:
yes on 7.x;   no failure on master (not in CI either)

**Applicable branches**:
only 7.x

**Failure history**:
14 failures over the last week

**Failure excerpt**:
```
org.elasticsearch.index.engine.ReadOnlyEngineTests > testSearcherId FAILED
    java.lang.AssertionError:
    Expected: not "0b4176a0054fce403abb72350d27706d167ec061d30031a49e5db37e4a50113b"
         but: was "0b4176a0054fce403abb72350d27706d167ec061d30031a49e5db37e4a50113b"
        at __randomizedtesting.SeedInfo.seed([5F23C74BF9BA02D2:BDA0039B9DAD7920]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.junit.Assert.assertThat(Assert.java:956)
        at org.junit.Assert.assertThat(Assert.java:923)
        at org.elasticsearch.index.engine.ReadOnlyEngineTests.testSearcherId(ReadOnlyEngineTests.java:373)
```

### PR DESCRIPTION
**Pull request by dnhatn**
The test fails when soft-deletes is disabled and deletes are processed out of order.

Closes #66869

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java [hunk 1]
```diff
@@ -345,37 +345,54 @@ public boolean keepFullyDeletedSegment(IOSupplier<CodecReader> readerIOSupplier)
...
+                        if (op instanceof Engine.Index) {
+                            engine.index((Engine.Index) op);
+                            if (latestOps.containsKey(op.id()) == false || latestOps.get(op.id()).seqNo() < op.seqNo()) {
+                                latestOps.put(op.id(), op);
+                                hasChanges = true;
+                            }
+                        } else if (op instanceof Engine.Delete) {
+                            engine.delete((Engine.Delete) op);
+                            final Engine.Operation lastOp = latestOps.get(op.id());
+                            hasChanges |= (lastOp instanceof Engine.Index && lastOp.seqNo() < op.seqNo());
```
Comment (henningandersen) [inline server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java:369]: I struggle to see how this line can have any effect? If we have an `Index` entry in the map, we must have already set `hasChanges = true` above?

**Review by henningandersen — COMMENTED** [review]
I have just one clarifying question, otherwise looking good.

File: server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java (hunk 1 above)
Comment (dnhatn) [reply server/src/test/java/org/elasticsearch/index/engine/ReadOnlyEngineTests.java:369]: This line avoids a situation where we have two deletes of the same document id. In that case `hasChanges` should stay intact.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by dnhatn** [comment]
Thanks Henning.

**Pull request merged**